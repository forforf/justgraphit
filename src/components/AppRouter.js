import React, {useState} from 'react';
import {HashRouter, Route} from 'react-router-dom';
import GraphObject from '../StorageModel/GraphObject';
import {AppBar} from '@mui/material';
import NavBar from './NavBar';
import Graphit from './Graphit';
import About from './About';
import Editor from './Editor';

function makePropless(Component, componentProps) {
  return (() => <Component {...componentProps} />)
}

function AppRouter({storage}) {
  const lastObject = storage.getLastStorageObject();

  // state
  const [graphObject, setGraphObject] = useState(GraphObject.fromPlainObject(lastObject));
  const [graphNameList, setGraphNameList] = useState(storage.getAllKeys());

  // CRUD-like handlers
  const changeSelectedGraph = graphName => {
    const newGraphData = storage.load(graphName);
    storage.updateHistory(graphName)
    const newGraphObject = new GraphObject(graphName, newGraphData)
    setGraphObject(newGraphObject);
  };

  const addNewNumber = (graphName, newNumberInput) => {
    const number = parseFloat(newNumberInput);
    const datetime = new Date().toISOString();
    const newEntry = { number, datetime };
    let newGraphData = [...graphObject.data, newEntry];
    // If we get a new graphName, create a new object
    if (graphName !== graphObject.name) {
      newGraphData = [newEntry];
      setGraphNameList([...graphNameList, graphName]);
    }
    const newGraphObject = new GraphObject(graphName, newGraphData);
    storage.save(graphName, newGraphData);
    setGraphObject(newGraphObject);
  }

  const deleteGraph = graphName => {
    if(window.confirm('Delete Graph?')) {
      storage.deleteObject(graphName);
      // Load last saved graph
      const lastName = storage.getLastKey()
      const lastData = storage.load(lastName);
      const newGraphObject = new GraphObject(lastName, lastData);
      setGraphObject(newGraphObject);
      setGraphNameList(storage.getAllKeys())
    }
  };

  // props
  const graphItProps = {
    graphObject,
    addNewNumber,
    changeSelectedGraph,
    deleteGraph,
    graphNameList
  }

  const editorProps = { graphObject, setGraphObject, saveToStorage: storage.save };

  return (
    <HashRouter>
      <AppBar title="JustGraphIt" />
      <NavBar />
      <Route exact path="/" component={makePropless(Graphit, {...graphItProps})} />
      <Route exact path="/about" component={About} />
      <Route exact path="/edit" component={makePropless(Editor, {...editorProps})} />
    </HashRouter>
  );
}

export default AppRouter
