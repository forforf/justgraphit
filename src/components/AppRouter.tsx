import React, { useState } from 'react';
import { HashRouter, Route } from 'react-router-dom';
import GraphObject from '../StorageModel/GraphObject';
import { AppBar } from '@mui/material';
import NavBar from './NavBar';
import Graphit, { GraphitProps } from './Graphit';
import About from './About';
import Editor, { EditorProps } from './Editor';
import {
  GraphEntryHandler,
  GraphHandler,
  ISODatetimeType,
  JustGraphitEntry,
} from '../JustGraphitTypes';
import Storage from '../StorageModel/Storage';

export type AppRouterProps = {
  storage: Storage;
};

function makePropless<T>(Component: React.ComponentType, componentProps: T) {
  function Propless(routerProps: unknown) {
    return <Component {...componentProps} {...routerProps} />;
  }

  Propless.displayName = `Propless(${Component.displayName || Component.name})`;
  return Propless;
}

function AppRouter({ storage }: AppRouterProps): JSX.Element {
  // const lastObject = storage.getLastStorageObject();
  const initialObject = storage.getInitialObject();

  // state
  const [graphObject, setGraphObject] = useState(
    GraphObject.fromPlainObject(initialObject)
  );
  const [graphNameList, setGraphNameList] = useState(storage.getAllKeys());

  // CRUD-like handlers
  const changeSelectedGraph: GraphHandler = (graphName) => {
    const newGraphData = storage.load(graphName);
    // storage.updateHistory(graphName)
    const newGraphObject = new GraphObject(graphName, newGraphData);
    setGraphObject(newGraphObject);
  };

  const addNewNumber: GraphEntryHandler = (graphName, newNumberInput) => {
    const number = parseFloat(newNumberInput);
    const datetime: ISODatetimeType =
      new Date().toISOString() as ISODatetimeType;
    const newEntry: JustGraphitEntry = { number, datetime };
    let newGraphData: JustGraphitEntry[] = [...graphObject.data, newEntry];
    // If we get a new graphName, create a new object
    if (graphName !== graphObject.name) {
      newGraphData = [newEntry];
      setGraphNameList([...graphNameList, graphName]);
    }
    const newGraphObject: GraphObject = new GraphObject(
      graphName,
      newGraphData
    );
    storage.save(graphName, newGraphData);
    setGraphObject(newGraphObject);
  };

  const deleteGraph: GraphHandler = (graphName) => {
    if (window.confirm('Delete Graph?')) {
      storage.deleteObject(graphName);
      // Load last saved graph
      const lastName = storage.getInitialKey();
      const lastData = storage.load(lastName);
      const newGraphObject = new GraphObject(lastName, lastData);
      setGraphObject(newGraphObject);
      setGraphNameList(storage.getAllKeys());
    }
  };

  // props
  const graphItProps = {
    graphObject,
    addNewNumber,
    changeSelectedGraph,
    deleteGraph,
    graphNameList,
  };

  const editorProps = {
    graphObject,
    setGraphObject,
    saveToStorage: storage.save,
  };

  return (
    <HashRouter>
      <AppBar title="JustGraphIt" />
      <NavBar />
      <Route
        exact
        path="/"
        component={makePropless<GraphitProps>(Graphit as React.ComponentType, {
          ...graphItProps,
        })}
      />
      <Route exact path="/about" component={About} />
      <Route
        exact
        path="/edit"
        component={makePropless<EditorProps>(Editor as React.ComponentType, {
          ...editorProps,
        })}
      />
    </HashRouter>
  );
}

export default AppRouter;
