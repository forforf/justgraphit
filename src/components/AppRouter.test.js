import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import AppRouter from './AppRouter';


let appRouterProps = null;
beforeEach(() => {
  const plainObj = {'empty graph': []};

  const mockStorage = {
    deleteObject: jest.fn(),
    getAllKeys: jest.fn(() => []),
    getLastKey: jest.fn(),
    getLastStorageObject: jest.fn(() => plainObj),
    load: jest.fn(),
    save: jest.fn()
  }

  appRouterProps = {
    storage: mockStorage
  }
})

test('deep snapshot', () => {
  const component = renderer.create(<AppRouter {...appRouterProps}/>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('shallow snapshot', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<AppRouter {...appRouterProps}/>);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot()
});
