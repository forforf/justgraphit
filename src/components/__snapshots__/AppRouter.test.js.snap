// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep snapshot 1`] = `
Array [
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
    title="JustGraphIt"
  />,
  <div
    className="MuiTabs-root css-1ujnqem-MuiTabs-root"
  >
    <div
      className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
      onScroll={[Function]}
      style={
        Object {
          "marginBottom": -0,
          "overflow": "hidden",
        }
      }
    >
      <div
        aria-orientation={null}
        className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
        onKeyDown={[Function]}
        role="tablist"
      >
        <a
          aria-selected={true}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          href="#/"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={0}
        >
          GraphIt
          <span
            className="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style={Object {}}
          />
        </a>
        <a
          aria-selected={false}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          href="#/edit"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Editor
        </a>
        <a
          aria-selected={false}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          href="#/about"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={-1}
        >
          About
        </a>
      </div>
    </div>
  </div>,
  <div
    className="Graphit"
  >
    <div
      className="wrapper"
    >
      <div
        className="row"
      >
        <div
          className="left col-2-3"
        >
          <div
            className="GraphInput"
          >
            <div
              className="GraphInput-name"
            >
              <h3>
                empty graph
              </h3>
            </div>
            <form
              onKeyPress={[Function]}
            >
              <div
                className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Name your graph"
                    required={false}
                    type="text"
                    value="empty graph"
                  />
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="css-1a1ihhj"
                    >
                      <span
                        className="notranslate"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                </div>
              </div>
              <br />
              <div
                className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Enter number to graph"
                    required={false}
                    type="number"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="css-1a1ihhj"
                    >
                      <span
                        className="notranslate"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                </div>
              </div>
              <br />
              <button
                className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-sghohy-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                Save
              </button>
            </form>
          </div>
          <div
            className="GraphChart"
          >
            <div>
              This is an EmptyGraph
            </div>
          </div>
        </div>
        <div
          className="right col-1-3"
        >
          <div
            className="GraphList"
          >
            <h3>
              Graph List
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`shallow snapshot 1`] = `
<HashRouter>
  <ForwardRef(AppBar)
    title="JustGraphIt"
  />
  <NavBar />
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/about"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/edit"
  />
</HashRouter>
`;
